src/dbami/helpers/apply_migrations.py

```diff
@@ -0,0 +1,117 @@
+import logging
+import time
+from typing import Literal, Optional, Sequence
+
+import asyncpg
+from buildpg import V, render
+
+from dbami.db import DB
+
+logger = logging.getLogger(__name__)
+
+
+async def grant_connect_privileges(conn: asyncpg.Connection, *role_names: str) -> None:
+    for role_name in role_names:
+        logger.warning("Granting connect privileges for role '%s'", role_name)
+        q, p = render(
+            """
+                DO $_$
+                    BEGIN
+                        EXECUTE FORMAT('GRANT CONNECT on database %s TO :role',
+                        CURRENT_DATABASE());
+                    END
+                $_$;
+            """,
+            role=V(role_name),
+        )
+        await conn.execute(q, *p)
+
+
+async def revoke_connect_privileges(conn: asyncpg.Connection, *role_names: str) -> None:
+    for role_name in role_names:
+        logger.warning("Revoking connect privileges for role '%s'", role_name)
+        q, p = render(
+            """
+                DO $_$
+                    BEGIN
+                        EXECUTE FORMAT('REVOKE CONNECT on database %s FROM :role',
+                        CURRENT_DATABASE());
+                    END
+                $_$;
+            """,
+            role=V(role_name),
+        )
+        await conn.execute(q, *p)
+
+
+async def active_connections_exist(conn: asyncpg.Connection) -> bool:
+    return bool(
+        await conn.fetchval(
+            """
+        SELECT EXISTS(
+            SELECT * FROM pg_stat_activity
+            WHERE
+                datname = current_database()
+                AND pid != pg_backend_pid()
+        )
+        """,
+        )
+    )
+
+
+async def wait_for_other_connections_to_close(
+    conn: asyncpg.Connection,
+    revoke_role_names: Sequence[str],
+    poll_interval_ms: int = 2000,
+) -> None:
+    if poll_interval_ms < 100:
+        logger.warning("Cowardly refusing to wait less than 100ms.")
+        poll_interval_ms = 100
+
+    await revoke_connect_privileges(conn, *revoke_role_names)
+
+    while await active_connections_exist(conn):
+        time.sleep(poll_interval_ms)
```

Comment:
time.sleep's parameter is seconds rather than milliseconds

Reply:
dividing by 1000